src/debug_sdfs_ed64.c

```diff
@@ -0,0 +1,598 @@
+	io_write(ED64_BASE_ADDRESS + ED64_REGISTER_SD_BASE + mode, 0xffff);
+	everdrive_sd_config = old_cfg;
+	// This seems necessary for everdrive somehow. If we don't try to set the
+	// bit length and restore, is not necessary.
+	wait_ticks(75);
+	io_write(ED64_BASE_ADDRESS + ED64_REGISTER_SD_STATUS, everdrive_sd_config);
+}
+
+void everdrive_sd_busy() {
+	while ((io_read(ED64_BASE_ADDRESS + ED64_REGISTER_SD_STATUS)
+		& ED64_SD_STATUS_BUSY) != 0);
+}
+
+void everdrive_sd_write_command(uint8_t val) {
+	everdrive_sd_set_mode(ED64_SD_MODE_CMD_WRITE);
+	io_write(ED64_BASE_ADDRESS + ED64_REGISTER_SD_BASE + ED64_SD_MODE_CMD_WRITE, val);
+	everdrive_sd_busy();
+}
+
+uint8_t everdrive_sd_read_command() {
+	everdrive_sd_set_mode(ED64_SD_MODE_CMD_READ);
+	// Even though this is exactly the same command as everdrive_sd_set_mode, it is
+	// required to actually read from the register.
+	io_write(ED64_BASE_ADDRESS + ED64_REGISTER_SD_BASE + ED64_SD_MODE_CMD_READ, 0xffff);
+	everdrive_sd_busy();
+	return io_read(ED64_BASE_ADDRESS + ED64_REGISTER_SD_BASE + ED64_SD_MODE_CMD_READ);
+}
+
+void everdrive_sd_write_data(uint8_t val) {
+	everdrive_sd_set_mode(ED64_SD_MODE_DATA_WRITE);
+	io_write(ED64_BASE_ADDRESS + ED64_REGISTER_SD_BASE + ED64_SD_MODE_DATA_WRITE,  0x00ff | (val << 8));
+}
+
+uint8_t everdrive_sd_read_data() {
+	everdrive_sd_set_mode(ED64_SD_MODE_DATA_READ);
+	// Even though this is exactly the same command as everdrive_sd_set_mode, it is
+	// required to actually write to the register.
+	io_write(ED64_BASE_ADDRESS + ED64_REGISTER_SD_BASE + ED64_SD_MODE_DATA_READ,  0xffff);
+	return io_read(ED64_BASE_ADDRESS + ED64_REGISTER_SD_BASE + ED64_SD_MODE_DATA_READ);
+}
+
+// Wait for and read the first byte
+static bool everdrive_sd_read_first(uint8_t res_buff[5]) {
+	uint32_t timeout = 64; // 8 bytes * 8 bits
+
+	set_everdrive_sd_bitlen(8);
+	res_buff[0] = everdrive_sd_read_command();
+
+	// Effectively we are bitshifting the command buffer until we find a zero
+	// (start bit), followed by another zero (transmission bit)
+	// We should be able to find it in 8 bytes
+	set_everdrive_sd_bitlen(1);
+	while ((res_buff[0] & 0xC0) != 0) {
+		if (!timeout--) return false;
+		res_buff[0] = everdrive_sd_read_command();
+	}
+	return true;
+}
+
+// Wait for and read an RX like response
+static bool everdrive_sd_read_response(uint8_t res_buff[5]) {
+	uint8_t timeout = 16;
+	if (!everdrive_sd_read_first(res_buff)) {
+		return false;
+	}
+
+	// read remaining bytes
+	set_everdrive_sd_bitlen(8);
+	res_buff[1] = everdrive_sd_read_command();
+	res_buff[2] = everdrive_sd_read_command();
+	res_buff[3] = everdrive_sd_read_command();
+	res_buff[4] = everdrive_sd_read_command();
+
+	// Make sure everything is consumed, we just don't use them
+	while(everdrive_sd_read_command() != 0xFF) {
+		if (!timeout--) {
+			return false;
+		}
+	};
+	return true;
+}
+
+static uint8_t everdrive_sd_resp_buff[5] = { 0 };
+static bool everdrive_sd_execute_command(uint8_t cmd, uint32_t arg) {
```

Comment:
Just like before, can't the response just be an argument to everdrive_sd_execute_command? I think I don't see a value in having a global `everdrive_sd_resp_buff` that can become stale or being accessed out-of-order. The response of a command should be a local variable IMO.

Reply:
Ah! Couldn't think of that, sure I will do this!